### PR TITLE
stricter hidden type wf-check

### LINKED ISSUE #114728: RPIT hidden types can be ill-formed
The following UB shouldn't compile:
```rust
type Static<'a> = &'static &'a ();
trait Extend<'a> {
    fn extend(self, _: &'a str) -> &'static str;
}
impl<'a> Extend<'a> for Static<'a> {
    fn extend(self, s: &'a str) -> &'static str {
        s
    }
}
fn boom<'a>(arg: Static<'_>) -> impl Extend<'a> {
    arg
}
fn main() {
    let y = boom(&&()).extend(&String::from("temporary"));
    println!("{}", y);
}
```
This regressed in v1.61. I guess it is #94081.

Discovered when investigating #114572.

### PR DESCRIPTION
**Pull request by aliemjay**
Combines #114740 and #114933 for a shared crater run, fixes #114728

#114740 should handle all the cases we unintentionally miss in #114933 and I expect #114933 to fix most of the crater regressions of #114740.

However this is still a regression:
```rust
trait Id<X> { type Ty; }
impl<X> Id<X> for () { type Ty = X; }
fn test<T>() -> impl Id<&'static T, Ty = impl Sized> {}
```
We should probably ignore the wf-check nested RPIT, similar to nested TAIT.

r? @ghost
cc @compiler-errors

### CODE REVIEW

**Comment by aliemjay** [comment]
(p=1 because it's only a small list)
@craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-114740/retry-regressed-list.txt

**Comment by oli-obk** [comment]
This PR fixes a soundness hole, where we were failing to check that hidden types are actually well formed wrt to lifetimes. From a not-well formed hidden type you can easily satisfy trait bounds that make no sense and thus allow treating any lifetime as another lifetime that lives longer.

The implementation adds well-formedness predicates, which should always be sound and at worst a performance issue (#114933 does the same thing). In addition, during wf check of RPIT and async fn return types we do region checking of these opaques within their owning function, catching the remaining soundness bugs (which is what #114740 does).

I do not believe there is any possible risk with this PR, beyond causing breakage for benign code that just forgot some bounds somewhere. Crater showed 3 such cases, which all have PRs opened, with only one not having been noticed by the crate owners yet.

@rfcbot fcp merge

File: tests/ui/traits/new-solver/auto-with-drop_tracking_mir.rs [hunk 1]
```diff
@@ -1,14 +1,16 @@
 // compile-flags: -Ztrait-solver=next -Zdrop-tracking-mir
 // edition: 2021
 // revisions: pass fail
-//[pass] check-pass
+//[pass] check-fail
+// WARN new-solver BUG.
```
Comment (lcnr) [inline tests/ui/traits/new-solver/auto-with-drop_tracking_mir.rs:5]: please instead check this to `known-bug` or `FIXME(-Ztrait-solver=next)`

**Comment by lcnr** [comment]
I force pushed to aliemjays repo as well by accident? :sweat_smile: alright, have to be more careful next time :sweat_smile:

**Pull request merged**